Remove DCOM support from *-Computer cmdlets

JamesWTruher opened the pull request:
Since DCOM is not supported in corefx there was a great deal of dead code in the computer cmdlets. This PR removes all vestiges of DCOM support from:

- Rename-Computer
- Restart-Computer
- Stop-Computer

removing about 4500 lines of dead code. Also, tests are updated to provide more complete coverage. 
I also removed test-connection completely to make way for @iSazonov upcoming PR
To improve coverage in tests, I created some test hook code which will test the cmdlet code without actually calling the WMI method to restart/rename/stop the system

JamesWTruher committed:
Add Tests for *-computer cmdlets

Remove all code protected by #!CORE
Add InternalTestHooks to improve testability of the cmdlets

JamesWTruher committed:
Add Tests for *-computer cmdlets

Remove all code protected by #!CORE
Add InternalTestHooks to improve testability of the cmdlets

JamesWTruher committed:
Remove all traces of DCOM from computer cmdlets

Reduce complexity in Test-Connection and update tests to reflect new behaviors

JamesWTruher committed:
Remove all traces of DCOM from computer cmdlets

Reduce complexity in Test-Connection and update tests to reflect new behaviors

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -669,213 +355,347 @@ private void ProcessWSManProtocolForTestConnection()
+        [ValidateSet(
+            "Default",
+            "Basic",
+            "Negotiate", // can be used with and without credential (without -> PSRP mapped to NegotiateWithImplicitCredential)
```
Comment should be updated: can be used with explicit or implicit credential

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -669,213 +355,347 @@ private void ProcessWSManProtocolForTestConnection()
+            "Negotiate", // can be used with and without credential (without -> PSRP mapped to NegotiateWithImplicitCredential)
+            "CredSSP",
+            "Digest",
+            "Kerberos")] // can be used with and without credential (not sure about implications)
```
Comment should be udpated: can be used with explicit or implicit credential

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -669,213 +355,347 @@ private void ProcessWSManProtocolForTestConnection()
+    $ret = $null
+    if ($null -eq $array[0])
+    {
+        $ret = Invoke-Command -ComputerName $computerName {$true} -SessionOption (New-PSSessionOption -NoMachineProfile) -ErrorAction SilentlyContinue
```
Should we update this script to use splatting so we don't have to repeat most of the parameters?

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -897,5380 +717,1211 @@ public void Dispose(bool disposing)
-        /// <param name="disposing"></param>
-        public void Dispose(bool disposing)
-        {
-            if (d
```
Update to same comment above

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -897,5380 +717,1211 @@ public void Dispose(bool disposing)
-        /// <param name="disposing"></param>
-        public void Dispose(bool disposing)
-        {
-            if (d
```
Update to same comment above

JamesWTruher replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -897,5380 +717,1211 @@ public void Dispose(bool disposing)
-        /// <param name="disposing"></param>
-        public void Dispose(bool disposing)
-        {
-            if (d
```
Define `5` as a const int with a meaningful name.  Or perhaps 0,1,5 should be an enum.

JamesWTruher replied:
I'll leave it as an int as that's really what it has to be turned into later (see https://msdn.microsoft.com/en-us/library/aa394058(v=vs.85).aspx)

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -0,0 +1,145 @@
+
+        It "Should support -computer parameter" {
+            Set-HookResult -result $defaultResultValue
+            Restart-Computer -computer (hostname) -ea Stop| Should BeNullOrEmpty
```
-computer takes an array, we should perhaps use: hostname, localhost

JamesWTruher replied:
fixed

SteveL-MSFT commented on test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1:
```diff
@@ -0,0 +1,74 @@
+
+        It "Should support -computer parameter" {
+            Set-HookResult -result $defaultResultValue
+            Stop-Computer -computer (hostname) -ea Stop| Should BeNullOrEmpty
```
-computer is an array, perhaps use: hostname, localhost

iSazonov commented:
I've already completely rewritten `Test-Connection` and  I'm working on tests. In the PR we can remove the cmdlet and its tests and I will push new PR in a few days.

JamesWTruher commented:
@iSazonov - I will remove the test-connection cmdlet and tests

JamesWTruher committed:
Remove Test-Connection from distribution

Improve usability of flag parameter for Restart/Stop Computer

JamesWTruher committed:
Remove Test-Connection from distribution

Improve usability of flag parameter for Restart/Stop Computer

SteveL-MSFT reviewed (approved):
My comments have been addressed

daxian-dbw commented on src/System.Management.Automation/engine/Types_Ps1Xml.cs:
```diff
@@ -236,6 +236,8 @@ public static IEnumerable<TypeData> Get()
             td24.Members.Add("IPV6Address",
                 new ScriptPropertyData(@"IPV6Address", GetScriptBlock(@"$iphost = [System.Net.Dns]::GetHostEntry($this.address)
           $iphost.AddressList | Where-Object { $_.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetworkV6 } | Select-Object -first 1"), null));
+            td24.Members.Add("Destination",
```
Since `Test-Connection` will be rewritten, do we still need this change?

iSazonov replied:
I don't use Win32_PingStatus - we can remove this here.

JamesWTruher replied:
removed

daxian-dbw commented on src/System.Management.Automation/engine/Types_Ps1Xml.cs:
```diff
@@ -1380,6 +1382,8 @@ public static IEnumerable<TypeData> Get()
                 new ScriptPropertyData(@"IPV6Address", GetScriptBlock(@"$iphost = [System.Net.Dns]::GetHostEntry($this.address)
           $iphost.AddressList | Where-Object { $_.AddressFamily -eq [System.Net.Sockets.AddressFamily]::InterNetworkV6 } | Select-Object -first 1"), null));
+            td167.Members.Add("Destination",
+                new AliasPropertyData("Destination","Address"));
```
Same here. Do we still need this change?

iSazonov replied:
I don't use Win32_PingStatus - we can remove this here.

JamesWTruher replied:
removed

daxian-dbw commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1481,5 +1489,15 @@ public static void SetTestHook(string property, bool value)
+            var fieldInfo = typeof(InternalTestHooks).GetField(property, BindingFlags.Static | BindingFlags.NonPublic);
+            if (fieldInfo != null)
+            {
+                fieldInfo.SetValue(null, value);
```
Maybe we should just change the existing method to take an `object` value. Boxing operation cannot be avoided anyway because `SetValue` method takes an object value. So it's better to not duplicate the code.

JamesWTruher commented on src/Microsoft.PowerShell.Commands.Management/commands/management/Computer.cs:
```diff
@@ -231,136 +148,29 @@ public Int32 ThrottleLimit
-
-            if(this.Protocol.Equals(ComputerWMIHelper.DcomProtocol , StringComparison.OrdinalIgnoreCase))
-            {
-                InvalidOperationException ex = new InvalidOperationException(ComputerResources.InvalidParameterDCOMNotSupported);
```
yes - done

JamesWTruher commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1466,6 +1466,14 @@ public static class InternalTestHooks
+        internal static bool TestStopComputer;
+        internal static bool TestWaitStopComputer;
+        internal static bool TestRenameComputer;
+        internal static int  TestStopComputerResults = 0;
```
fixed

JamesWTruher commented on src/System.Management.Automation/engine/Utils.cs:
```diff
@@ -1466,6 +1466,14 @@ public static class InternalTestHooks
+        internal static bool TestWaitStopComputer;
+        internal static bool TestRenameComputer;
+        internal static int  TestStopComputerResults = 0;
+        internal static int  TestRenameComputerResults = 0;
```
fixed

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Rename-Computer.Tests.ps1:
```diff
@@ -0,0 +1,97 @@
+    ${TesthooksType}::SetTestHook($TesthookResultName, $result)
+}
+
+# protect the user from from stop their computer 
```
fixed

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Rename-Computer.Tests.ps1:
```diff
@@ -0,0 +1,97 @@
+        It "Should produce a reboot warning when renaming computer" {
+            Set-HookResult -result $defaultResultValue
+            $newname = "mynewname"
+            $result = Rename-Computer -ea Stop -ComputerName . -NewName "$newname" -Pass -WarningAction SilentlyContinue -WarningVariable WarnVar
```
fixed

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -0,0 +1,145 @@
+
+        It "Should restart the local computer" {
+            Set-HookResult -result $defaultResultValue
+            Restart-Computer -ea Stop| Should BeNullOrEmpty
```
fixed

JamesWTruher committed:
Refactor testhook setting code to module

remove unneeded DCOM resources

JamesWTruher committed:
Refactor testhook setting code to module

remove unneeded DCOM resources

JamesWTruher commented:
@daxian-dbw @adityapatwardhan I think I've got your feedback covered. Please let me know if you have more comments

SteveL-MSFT committed:
Put command discovery before scripts for Unix (#5116)

* put command discovery before scripts for Unix

* remove unnecessary test

JamesWTruher committed:
Unify cmdlets with parameter 'Encoding' to be of type System.Text.Encoding (#5080)

This unifies file encoding across the inbox cmdlets to be UTF-8 without a BOM for all platforms. This is a breaking change as cmdlets on windows have a number of different encodings. This supports better interoperability with tradition Linux shells as we are using the same encoding.

Validate that files are created with UTF-8 encoding without BOM
Update tests to validate Encoding parameter to new type and create new tests for
parameter type validation.

[Breaking Change] The '-Encoding Byte' has been removed from the filesystem provider cmdlets. A new parameter '-AsByteStream' is now added to indicate that a byte stream is required as input, or output will be a stream of bytes.

daxian-dbw committed:
Fix appimage package name to be lower case 'powershell-..' (#5206)

joandrsn committed:
Added BinPath, Description, UserName and Delayed Auto Start to Get-Service (#4907)

Added the following to the  Get-Service  command:
•BinaryPathName (Was suggested as  ServicePath )
•Description
•UserName (Was suggested as  LogOnAs )
•DelayedAutoStart
•ServiceStartupType

The ServiceStartupType was also added, providing the user access to use services like in  services.msc , having  Automatic ,  Automatic (Delayed Start) ,  Manual  and  Disabled  as options (Also includes  InvalidValue  as an initial value for  ServiceStartupType )

daxian-dbw committed:
Update installation doc and README.md to include the powershell binary archive packages (#5212)

* Update README.md

* Update docs for powershell binary archive packages

* Update the chart

adityapatwardhan committed:
Update docs and scripts for beta.9 (#5217)

daxian-dbw committed:
Update ChangeLog for beta.9 release (#5216)

daxian-dbw committed:
Update install/build scripts and dockerfiles (#5222)

bergmeister committed:
Add -AsHashtable switch to ConvertFrom-Json for issue #3623 (#5043)

Address #3623.

Implementation symmetric to traditional code path that returns a PSCustomObject. Code is shared on the top level but 2 low level methods were difficult to share, therefore 2 separate but similar methods were created for HashTables.
All existing tests related to this change were adapted to also test against this new switch and were slightly improved.

-AsHashtable is an existing pattern used in Group-Object

daxian-dbw committed:
Merge beta.9 release branch to master

bergmeister committed:
Fix 'demos/SSHRemoting/README.md' using the new name 'pwsh' (#5236)

vors committed:
Remove @vors from engine area owners. Add the Language area. (#5223)

TravisEz13 committed:
Add term rules  (#5213)

* Add Terms Rules

* remove offensive term

* Add code owner for compliance files

Marusyk committed:
Export-Csv Test Improvements (#5150)

SteveL-MSFT committed:
-Verbose should not override $ErrorActionPreference (#5113)

* -Verbose should not override $ErrorActionPreference

* apply same fix to -debug and added test case

* address Aditya's feedback

adityapatwardhan commented:
@SteveL-MSFT @JamesWTruher How would test hooks work when tests are executed in parallel?

SteveL-MSFT committed:
Use consistent '(c)' for copyright symbol (#5210)

- Remove the year about copyright
- Fix casing of `All rights reserved`
- Replace Unicode characters representing single quote with a single quote

SteveL-MSFT committed:
Fix 'get-item -literalpath a*b' to return error if `a*b` doesn't actually exist (#5197)

SteveL-MSFT committed:
Fix 'Import-module' to not report a loaded module was not found (#5238)

Missed a 'break' statement after the module was loaded successfully, so the loop continues when it shouldn't.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+        It "Should support WsmanAuthentication types" {
+            $authChoices = "Default","Basic","Negotiate","CredSSP","Digest","Kerberos"
+            foreach ( $auth in $authChoices ) {
+                Restart-Computer -WsmanAuthentication $auth | Should BeNullOrEmpty
```
since this is a straight shot into the cim session creation, there isn't a lot to be gained here, but for coverage of the parameter setting I added this test. There doesn't seem to be any property on the cimsession which describes how the session was created. I could pull all but one which would provide the coverage.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+                { Restart-Computer -timeout 3 -ErrorAction Stop } | ShouldBeErrorId "RestartComputerInvalidParameter,Microsoft.PowerShell.Commands.RestartComputerCommand"
+            }
+        }
+    }
```
no more -asjob. It wasn't supported in Core

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1:
```diff
@@ -0,0 +1,59 @@
+
+        It "Should support -Computer parameter" {
+            Set-TesthookResult -testhookName $stopTesthookResultName -Value $defaultResultValue
+            $computerNames = "localhost",(hostname)
```
fixed

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1:
```diff
@@ -0,0 +1,59 @@
+        It "Should support WsmanAuthentication types" {
+            $authChoices = "Default","Basic","Negotiate","CredSSP","Digest","Kerberos"
+            foreach ( $auth in $authChoices ) {
+                Stop-Computer -WsmanAuthentication $auth | Should BeNullOrEmpty
```
same as above

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Stop-Computer.Tests.ps1:
```diff
@@ -0,0 +1,59 @@
+
+        }
+    }
+
```
no `-asjob` parameter now

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Rename-Computer.Tests.ps1:
```diff
@@ -0,0 +1,80 @@
+try 
+{
+    # set up for testing
+    $PSDefaultParameterValues["it:skip"] = ! $IsWindows
```
we've seen a number of issues with this guidance, and I think it needs revision. What I have seems less error prone

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Management/Restart-Computer.Tests.ps1:
```diff
@@ -0,0 +1,100 @@
+
+        It "Should support -computer parameter" {
+            Set-TesthookResult -testhookName $restartTesthookResultName -value $defaultResultValue
+            $computerNames = "localhost",(hostname)
```
fixed

SteveL-MSFT committed:
use rcedit to embed icon and version information into pwsh.exe (#5178)

pwsh.exe today doesn't contain file version information and the icon is only associated with the shortcut file and not the exe

Fix is to use rcedit to embed:

icon
product version
file version
product name
copyright
Fix #2883
Fix #5166
Fix #5034

TravisEz13 committed:
Make sure web hook gets called (#5249)

if something goes wrong during the build, we might exit and not update the badge or fire the webhook.

Enable the after_success and after_failure stages of the travis-ci build so we can have a better chance of getting these done.

TravisEz13 committed:
update docker file for nanoserver 1709 release (#5252)

JamesWTruher committed:
fix merge conflict

adityapatwardhan commented:
@JamesWTruher The PR has many unrelated commits. Can you rebase?

JamesWTruher commented:
something happened when I attempted to correct a merge conflict. I'm closing this PR and opening a new one as that is going to be easier than trying to fix this.
